src/transaction.js

```diff
@@ -170,9 +229,100 @@ var VALUE_UINT64_MAX = new Buffer('ffffffffffffffff', 'hex')
  * hashType, and then hashes the result.
  * This hash can then be used to sign the provided transaction input.
  */
-Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashType) {
-  typeforce(types.tuple(types.UInt32, types.Buffer, /* types.UInt8 */ types.Number), arguments)
+Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashType, segWit, amount) {
+  typeforce(types.tuple(types.UInt32, types.Buffer, /* types.UInt8 */ types.Number, types.maybe(types.Boolean)), arguments)
+
+  if (segWit) {
+    return this.hashForSignatureV2(inIndex, prevOutScript, amount, hashType)
+  } else {
+    return this.hashForSignatureV1(inIndex, prevOutScript, hashType)
+  }
+}
+
+Transaction.prototype.hashForSignatureV2 = function (inIndex, prevOutScript, amount, hashType) {
```

Comment:
Not sure if this was cleared up - segwit has a different signature hashing scheme which commits the input amount in the signature, amongst other changes. It has a reusable midstate for all inputs in a transaction which may be cached, so avoids the O(n^2) hashing issue.